mathphys/imgproc.py

```diff
@@ -0,0 +1,1350 @@
+
+        return res
+
+    def _update_image_roi(self, roi=None, *args, **kwargs):
+        """."""
+        super()._update_image_roi(roi=roi)
+
+        # fit roi
+        param, roi_fit, roi_error = \
+            self._fitgauss.calc_fit(
+                self, self.roi_proj, self.roi_indcs, self.roi_center)
+        self._roi_sigma, self._roi_mean, self._roi_amp, _ = param
+        self._roi_fit, self._roi_fit_error = roi_fit, roi_error
+
+
+class Image2D_Fit(Image2D):
+    """2D Image Fit."""
+
+    def __init__(self, roix=None, roiy=None, fitgauss=None, *args, **kwargs):
+        """."""
+        # benchmark for sizes=(1024, 1280)
+        #   2.7 ms ± 23.2 µs per loop
+        #   (mean ± std. dev. of 7 runs, 100 loops each)
+        self._fitx = None
+        self._fity = None
+        self._angle = 0
+        self._fitgauss = fitgauss or FitGaussian
+        super().__init__(*args, **kwargs)
+        self._update_image_fit(roix=roix, roiy=roiy)
+
+    @property
+    def fity(self):
+        """."""
+        return self._fity
+
+    @property
+    def fitx(self):
+        """."""
+        return self._fitx
+
+    @property
+    def roiy(self):
+        """."""
+        return self.fity.roi
+
+    @roiy.setter
+    def roiy(self, value):
+        """."""
+        self._update_image_fit(roix=self.fitx.roi, roiy=value)
+
+    @property
+    def roix(self):
+        """."""
+        return self.fitx.roi
+
+    @roix.setter
+    def roix(self, value):
+        """."""
+        self._update_image_fit(roix=value, roiy=self.fity.roi)
+
+    @property
+    def roi(self):
+        """."""
+        return self.fitx.roi, self.fity.roi
+
+    @roi.setter
+    def roi(self, value):
+        """."""
+        self._update_image_fit(*value)
+
+    @property
+    def angle(self):
+        """Image tilt angles [deg]."""
+        return self._angle
+
+    def update_roi_with_fwhm(self, fwhmx_factor=2, fwhmy_factor=2):
+        """."""
+        self.fitx.update_roi_with_fwhm(fwhm_factor=fwhmx_factor)
+        self.fity.update_roi_with_fwhm(fwhm_factor=fwhmy_factor)
+
+    def calc_angle_with_roi(self):
+        """Calculate image tilt angle within ROI.
+
+        A linear y = ax + b is performed over the image and the angle is
+        taken from the arctan of the angular coefficient. Each image point
+        is weighted by the fourth power of the image intensity.
+        """
+        roix, roiy = self.fitx.roi, self.fity.roi
+        indcsx, indcsy = self.fitx.roi_indcs, self.fity.roi_indcs
+        mx, my = _np.meshgrid(indcsx, indcsy)
+        data = self.data[slice(*roiy), slice(*roix)]
+        data = data * data
+        data = data * data
+        mxd = mx * data
+        a11 = _np.sum(mx * mxd)
+        a12 = _np.sum(mxd)
+        a22 = _np.sum(data)
+        b1 = _np.sum(my * mxd)
+        b2 = _np.sum(my * data)
+        a = _np.array([[a11, a12], [a12, a22]])
+        b = _np.array([b1, b2])
+        v = _np.linalg.solve(a, b)
```

Comment:
another way of implementing this fitting would be:
```
coef, *_ = np.polynomial.polynomial.polyfit(mx.ravel(), my.ravel(), deg=1, w=data.ravel(), rcond=None)
```

I think this way is clearer for the reader, but I don't know how it compares in terms of performance.

Reply:
Actually, to get the same power of the intensity as you used, the weights you should pass to the `numpy`'s method is the intensity square, because it will square the weights in the minimization process, resulting in a total weight of intensity to the fourth power.

Reply:
way too slower. 3.4 ms against 0.8 ms